library/alloc/src/collections/vec_deque.rs

```diff
@@ -1083,24 +1082,18 @@ impl<T> VecDeque<T> {
         self.tail == self.head
     }
 
-    fn range_start_end<R>(&self, range: R) -> (usize, usize)
+    fn range_tail_head<R>(&self, range: R) -> (usize, usize)
     where
         R: RangeBounds<usize>,
     {
-        let len = self.len();
-        let start = match range.start_bound() {
-            Included(&n) => n,
-            Excluded(&n) => n + 1,
-            Unbounded => 0,
-        };
-        let end = match range.end_bound() {
-            Included(&n) => n + 1,
-            Excluded(&n) => n,
-            Unbounded => len,
-        };
-        assert!(start <= end, "lower bound was too large");
-        assert!(end <= len, "upper bound was too large");
-        (start, end)
+        // SAFETY: This buffer is only used to check the range. It might be partially
+        // uninitialized, but `check_range` needs a contiguous slice.
+        // https://github.com/rust-lang/rust/pull/75207#discussion_r471193682
```

Comment:
Creating a partially uninitialized slice is UB. So I don't think the safety comment here is accurate.

Reply:
I thought this was something the standard library could assume. Wouldn't that make `buffer_as_slice` UB too?

Reply:
If this was meant to rely on "the standard library is special and can bend the rules", the comment should explicitly say so. The standard library can't just implicitly do this, the risk is too high that someone will copy this code and use it in their own crate.

However, even then it is better to avoid; usually we only do that for layout assumptions (Cc https://github.com/rust-lang/unsafe-code-guidelines/issues/90). For other kinds of UB, it is typically considered a bug.

> Wouldn't that make buffer_as_slice UB too?

Maybe? I am not familiar with the code I am afraid, I just read the safety comment.

Reply:
Yeah, looking at the code, `buffer_as_slice` should likely return a raw slice instead of a reference.

Reply:
> If this was meant to rely on "the standard library is special and can bend the rules", the comment should explicitly say so.

You're right. My mistake.

> However, even then it is better to avoid; usually we only do that for layout assumptions

That makes sense. I was mostly using `buffer_as_slice` as a reference, but I agree that avoiding that assumption is better.

Reply:
Yeah, there's lots of old code here that was written before [we had clear rules](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), and so does not follow today's rules. This case here is not a big deal because the discussion is still open whether we want this to be UB or not (that's https://github.com/rust-lang/unsafe-code-guidelines/issues/90), but while we discuss it's better to [follow the rules](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), even in the standard library, also to learn how hard/annoying it is to work with these rules.

Reply:
Oh also the standard library currently does a [poor job](https://github.com/rust-lang/unsafe-code-guidelines/issues/90) at documenting when it relies on such assumptions, but I am pushing for new code that we add to do better. :)

Reply:
That all sounds great, :)